core/new-gui/src/app/dashboard/component/top-bar/user-icon/user-login/user-login-modal.component.ts

```diff
@@ -1,55 +1,51 @@
 import { Component, OnInit } from "@angular/core";
-import { NgbActiveModal } from "@ng-bootstrap/ng-bootstrap";
 import { UserService } from "../../../../../common/service/user/user.service";
 import { FormBuilder, FormControl, FormGroup, Validators } from "@angular/forms";
 import { isDefined } from "../../../../../common/util/predicate";
 import { filter } from "rxjs/operators";
 import { UntilDestroy, untilDestroyed } from "@ngneat/until-destroy";
+import { NzModalRef } from "ng-zorro-antd/modal";
 
 /**
- * NgbdModalUserLoginComponent is the pop up for user login/registration
- *
- * @author Adam
+ * UserLoginModalComponent is the pop up for user login/registration
  */
 @UntilDestroy()
 @Component({
-  selector: "texera-ngbdmodal-user-login",
-  templateUrl: "./ngbdmodal-user-login.component.html",
-  styleUrls: ["./ngbdmodal-user-login.component.scss"],
+  selector: "texera-user-login-modal",
+  templateUrl: "./user-login-modal.component.html",
+  styleUrls: ["./user-login-modal.component.scss"],
 })
-export class NgbdModalUserLoginComponent implements OnInit {
-  public selectedTab = 0;
+export class UserLoginModalComponent implements OnInit {
   public loginErrorMessage: string | undefined;
   public registerErrorMessage: string | undefined;
   public allForms: FormGroup;
 
-  constructor(private formBuilder: FormBuilder, public activeModal: NgbActiveModal, private userService: UserService) {
+  constructor(private formBuilder: FormBuilder, public modal: NzModalRef, private userService: UserService) {
     this.allForms = this.formBuilder.group({
       loginUsername: new FormControl("", [Validators.required]),
       registerUsername: new FormControl("", [Validators.required]),
-      loginPassword: new FormControl("", [Validators.required]),
-      registerPassword: new FormControl("", [Validators.required]),
-      registerConfirmationPassword: new FormControl("", [Validators.required]),
+      loginPassword: new FormControl("", [Validators.required, Validators.minLength(6)]),
+      registerPassword: new FormControl("", [Validators.required, Validators.minLength(6)]),
+      registerConfirmationPassword: new FormControl("", [Validators.required, this.confirmationValidator]),
     });
   }
 
   ngOnInit() {
     this.detectUserChange();
   }
 
-  public errorMessageUsernameNull(): string {
-    return "Username required";
+  public updateConfirmValidator(): void {
+    // immediately update validator (asynchronously to wait for value to refresh)
+    Promise.resolve().then(() => this.allForms.controls.registerConfirmationPassword.updateValueAndValidity());
```

Comment:
shouldn't we avoid using Promise?

Reply:
basically here we cannot run that line synchronously, because at this moment the view is not refreshed by angular yet. We want to run it asynchronously after the view is refreshed (but immediately). To do this we can either do `setTimeout(0)` or `Promise.resolve()` , both will schedule the callback to run immediately. Either way is fine.

Reply:
changed to `setTimeout(0)` because another place with a similar logic also uses `setTimeout(0)`